capellambse/decl.py

```diff
@@ -0,0 +1,404 @@
+    reliable way to know how much of the YAML input has been consumed.
+    It is therefore advised to reload or discard the model immediately
+    in these cases, to avoid working with an inconsistent state.
+
+    Even though the YAML layout suggests linear execution from top to
+    bottom, the actual order in which modifications are executed is
+    implementation defined. This is necessary to support promises with
+    ``!promise``, but reorderings are still possible even if no promises
+    are used in an input document.
+    """
+    instructions = collections.deque(load(file))
+    promises = dict[Promise, capellambse.ModelObject]()
+    deferred = collections.defaultdict[Promise, list[_FutureAction]](list)
+
+    while instructions:
+        instruction = instructions.popleft()
+        parent = instruction.pop("parent")
+        if isinstance(parent, UUIDReference):
+            parent = model.by_uuid(parent.uuid)
+        for op_type, apply_op in _OPERATIONS.items():
+            try:
+                op = instruction.pop(op_type)
+            except KeyError:
+                continue
+            for promise, outcome in apply_op(promises, parent, op):
+                if isinstance(outcome, dict):
+                    deferred[promise].append(outcome)
+                else:
+                    if promise in promises:
+                        raise ValueError(
+                            f"promise_id defined twice: {promise.identifier}"
+                        )
+                    promises[promise] = outcome
+                    instructions.extend(deferred.pop(promise, ()))
+        if instruction:
+            keys = ", ".join(instruction)
+            raise ValueError(f"Unrecognized keys in instruction: {keys}")
+    if deferred:
+        raise UnfulfilledPromisesError(frozenset(deferred))
+
+
+def _operate_create(
+    promises: dict[Promise, capellambse.ModelObject],
+    parent: capellambse.ModelObject,
+    creations: dict[str, t.Any],
+) -> cabc.Generator[_OperatorResult, t.Any, None]:
+    for attr, value in creations.items():
+        if not isinstance(value, cabc.Iterable):
+            raise TypeError("values below `create:*:` must be lists")
+
+        yield from _create_complex_objects(promises, parent, attr, value)
+
+
+def _operate_delete(
+    promises: dict[Promise, capellambse.ModelObject],
+    parent: capellambse.ModelObject,
+    deletions: dict[str, t.Any],
+) -> cabc.Iterable[_OperatorResult]:
+    del promises
+
+    for attr, objs in deletions.items():
+        try:
+            target = getattr(parent, attr)
+        except AttributeError:
+            raise TypeError(
+                "Cannot delete object:"
+                f" {type(parent).__name__} has no attribute {attr!r}"
+            ) from None
+        if not isinstance(target, common.ElementList) or not isinstance(
+            objs, list
+        ):
+            delattr(parent, attr)
```

Comment:
Does this `delattr` call modify the XML code? If so how? I expect some hidden execution on the `Accessor`. Also [this](https://stackoverflow.com/a/1121068/5628391) was an interesting read for me.

Reply:
Looking closely, this bit was actually missed by the unit tests.

I've added one, and can now confidently answer:

> Does this delattr call modify the XML code?

Yes, it does! :)

> If so how?

[The same way getattr/setattr-like statements do.](https://docs.python.org/3/reference/datamodel.html#implementing-descriptors)